70_FLIPFluids/app_resources/compute/radix_sort/radixSort.comp.hlsl

```diff
@@ -0,0 +1,205 @@
+
+    // reduce, downsweep step
+    for (uint i = idx; i < NumSortBins * numSubgroups; i += WorkgroupSize)
+    {
+        uint v = localHistogram[i];
+        uint sum = WaveActiveSum(v);
+        uint prefixSum = WavePrefixSum(v);
+        localHistogram[i] = prefixSum;
+        if (ls_id == 0)
+            histogramSums[i / subgroupSize] = sum;
+    }
+    GroupMemoryBarrierWithGroupSync();
+
+    uint reduceOffset = NumSortBins * numSubgroups / subgroupSize;
+    if (idx < reduceOffset)
+    {
+        uint v = histogramSums[idx];
+        uint sum = WaveActiveSum(v);
+        uint prefixSum = WavePrefixSum(v);
+        histogramSums[idx] = prefixSum;
+        if (ls_id == 0)
+            histogramSums[reduceOffset + idx / subgroupSize] = sum;
+    }
+    GroupMemoryBarrierWithGroupSync();
+
+    uint reduceSize = NumSortBins * numSubgroups / subgroupSize / subgroupSize;
+    if (idx < reduceSize)
+    {
+        uint v = histogramSums[reduceOffset + idx];
+        uint prefixSum = WavePrefixSum(v);
+        histogramSums[reduceOffset + idx] = prefixSum;
+    }
+    GroupMemoryBarrierWithGroupSync();
+
+    if (idx < reduceOffset)
+        histogramSums[idx] += histogramSums[reduceOffset + idx / subgroupSize];
+    GroupMemoryBarrierWithGroupSync();
+
+    for (uint i = idx; i < NumSortBins * numSubgroups; i += WorkgroupSize)
+        localHistogram[i] += histogramSums[i / subgroupSize];
+    GroupMemoryBarrierWithGroupSync();
+
+    // scatter step
+    for (uint i = 0; i < NumPartitions; i++)
+    {
+        uint bin = bins[i];
+        offsets[i] += localHistogram[numSubgroups * bin + s_id];
+        GroupMemoryBarrierWithGroupSync();
+
+        if (subgroupHist[i] > 0)
+            InterlockedAdd(localHistogram[numSubgroups * bin + s_id], subgroupHist[i]);
+        GroupMemoryBarrierWithGroupSync();
+    }
+
+    if (idx < NumSortBins)
+    {
+        uint v = idx == 0 ? 0 : localHistogram[numSubgroups * idx - 1];
+        uint passIdx = params.bitShift / 8;
+        histogramSums[idx] = globalHistograms[NumSortBins * passIdx + idx] + partitionHistogram[NumSortBins * partitionIdx + idx] - v;
+    }
+    GroupMemoryBarrierWithGroupSync();
+
+    for (uint i = 0; i < NumPartitions; i++)
+    {
+        localHistogram[offsets[i]] = keys[i];
+    }
+    GroupMemoryBarrierWithGroupSync();
+
+    // binning
+    for (uint i = idx; i < PartitionSize; i += WorkgroupSize)
+    {
+        uint key = localHistogram[i];
+        uint bin = glsl::bitfieldExtract(key, params.bitShift, 8);
+        uint dstOffset = histogramSums[bin] + i;
+        if (dstOffset < params.numElements)
+            setKey(outputBuffer, dstOffset, key);
+
+#ifdef USE_KV_PAIRS
+        keys[i / WorkgroupSize] = dstOffset;
+#endif
+    }
+
+#ifdef USE_KV_PAIRS
+    GroupMemoryBarrierWithGroupSync();
+
+    for (uint i = 0; i < NumPartitions; i++)
+    {
+        localHistogram[offsets[i]] = values[i];
+    }
+    GroupMemoryBarrierWithGroupSync();
```

Comment:
all of this is already available with `workgroup::inclusive` as used in Example 24. Arithmetic test and counting sort example